src/timers.rs

```diff
@@ -0,0 +1,218 @@
+                /// constructee guarantees to never modify the timer channel.
+                unsafe fn new() -> Self {
+                    Self {}
+                }
+
+                /// Allow the channel to generate DMA requests.
+                #[allow(dead_code)]
+                pub fn listen_dma(&self) {
+                    let regs = unsafe { &*<$TY>::ptr() };
+                    regs.dier.modify(|_, w| w.[< cc $index de >]().set_bit());
+                }
+
+                /// Operate the channel as an output-compare.
+                ///
+                /// # Args
+                /// * `value` - The value to compare the sampling timer's counter against.
+                #[allow(dead_code)]
+                pub fn to_output_compare(&self, value: u32) {
+                    let regs = unsafe { &*<$TY>::ptr() };
+                    assert!(value <= regs.arr.read().bits());
+                    regs.[< ccr $index >].write(|w| w.ccr().bits(value));
+                    regs.[< $ccmrx _output >]()
+                        .modify(|_, w| unsafe { w.[< cc $index s >]().bits(0) });
+                }
+
+                /// Operate the channel in input-capture mode.
+                ///
+                /// # Args
+                /// * `input` - The input source for the input capture event.
+                #[allow(dead_code)]
+                pub fn into_input_capture(self, input: hal::stm32::tim2::[< $ccmrx _input >]::[< CC $index S_A >]) -> [< Channel $index InputCapture >]{
+                    let regs = unsafe { &*<$TY>::ptr() };
+                    regs.[< $ccmrx _input >]().modify(|_, w| w.[< cc $index s>]().variant(input));
+
+                    [< Channel $index InputCapture >] {}
+                }
+            }
+
+            impl [< Channel $index InputCapture >] {
+                /// Get the latest capture from the channel.
+                #[allow(dead_code)]
+                pub fn latest_capture(&mut self) -> Result<Option<u32>, ()> {
+                    // Note(unsafe): This channel owns all access to the specific timer channel.
+                    // Only atomic operations on completed on the timer registers.
+                    let regs = unsafe { &*<$TY>::ptr() };
+                    let sr = regs.sr.read();
+                    let ccx = regs.[< ccr $index >].read();
+
+                    let result = if sr.[< cc $index if >]().bit_is_set() {
+                        regs.sr.modify(|_, w| w.[< cc $index if >]().clear_bit());
+                        Some(ccx.ccr().bits())
+                    } else {
+                        None
+                    };
+
+                    // If there is an overcapture, return an error.
+                    if sr.[< cc $index of >]().bit_is_clear() {
+                        Ok(result)
+                    } else {
+                        Err(())
+                    }
+                }
```

Comment:
I think we might want to read the SR twice, like so:
```Rust                                        
let sr1 = regs.sr.read(); // Read SR to check if CCxIF is set                   
let ccx = regs.[< ccr $index >].read();                                         
let sr2 = regs.sr.read(); // Read SR again to make sure CCxOF was not set       
                                                                                
let result = if sr1.[< cc $index if >]().bit_is_set() {                                          
    Some(ccx.ccr().bits())                                                         
} else {                                                                        
    None                                                                        
};                                                                              
                                                                                
// If there is an overcapture, return an error.                                 
if sr2.[< cc $index of >]().bit_is_clear() {                                    
    Ok(result)                                                                  
} else {                                                                        
    Err(())                                                                     
}  
```

I think we want to check the CCxIF bit to make sure the capture register has been updated before we read the capture register. But then we then want to check the CCxOF bit after we have read the capture register to make sure an overcapture didn't happen, which could have invalidated our read of the capture register.

According to the reference manual, we don't need to explicitly clear the CCxIF flag, this happens automatically when we read the capture register.

However, what about the scenario where the CCxIF flag is set, therefore the capture register is valid, we read it, but then another capture comes in (basically we just barely read in time before an overcapture occurred)? Do we want to detect and handle that case? Should we allow for returning two timestamps, or just treat it as if an overcapture occurred, as that likely indicates a timing problem in the code?

Should we clear the CCxOF bit when an overcapture occurs to return things to a good state?

Reply:
> However, what about the scenario where the CCxIF flag is set, therefore the capture register is valid, we read it, but then another capture comes in (basically we just barely read in time before an overcapture occurred)? Do we want to detect and handle that case? Should we allow for returning two timestamps, or just treat it as if an overcapture occurred, as that likely indicates a timing problem in the code?

I don't think that's indicative of ill-operating code. In fact, I think that would indicate perfectly functional code. As long as the capture is never lost, there's no issue with the data. We will get around to the timestamp in the future.

> Should we clear the CCxOF bit when an overcapture occurs to return things to a good state?

It doesn't matter right now (since we panic if there is ever an error), but you're right. I've updated it to clear the flag before returning the error.